miqa/core/rest/permissions.py

```diff
@@ -1,12 +1,58 @@
 from django.contrib.auth.models import User
 from django.db.models import Model
+from django.shortcuts import get_object_or_404
+from django.utils.functional import wraps
+from guardian.shortcuts import get_perms
 from rest_framework import status
 from rest_framework.exceptions import APIException
 from rest_framework.permissions import SAFE_METHODS, BasePermission
 from rest_framework.request import Request
+from rest_framework.response import Response
 from rest_framework.views import View
 
-from miqa.core.models import Experiment
+from miqa.core.models import Experiment, Project
+
+
+def project_permission_required(
+    read_access=True, review_access=False, superuser_access=False, **kwargs
+):
+    def decorator(view_func):
+        def _wrapped_view(viewset, *args, **kwargs):
+            lookup_mapping = {
+                'ProjectViewSet': 'pk',
+                'ExperimentViewSet': 'experiments__pk',
+                'ScanViewSet': 'experiments__scans__pk',
+                'ScanDecisionViewSet': 'experiments__scans__decisions__pk',
+                'FrameViewSet': 'experiments__scans__frames__pk',
+            }
+            lookup_dict = {lookup_mapping[viewset.__class__.__name__]: kwargs['pk']}
+            project = get_object_or_404(Project, **lookup_dict)
+
+            user = viewset.request.user
+            user_perms_on_project = get_perms(user, project)
+            has_review_perm = all(
+                [
```

Comment:
You can remove the square braces; `all` is one of the special built-ins that can accept a generator expression syntax.

Reply:
Oh yeah, you told me that a few PRs ago and I keep forgetting!